vendor/brave-ios/Ledger/BATBraveLedger.mm

```diff
@@ -190,9 +193,129 @@ - (void)dealloc
+        switch (self.migrationType) {
+          case BATLedgerDatabaseMigrationTypeDefault:
+            BLOG(ledger::LogLevel::LOG_ERROR) << "DB: Full migration failed, attempting BAT only migration." << std::endl;
+            self.dataMigrationFailed = YES;
+            self.migrationType = BATLedgerDatabaseMigrationTypeTokensOnly;
+            [self resetRewardsDatabase];
+            // attempt re-initialize without other data
+            [self initializeLedgerService:YES];
+            return;
+          case BATLedgerDatabaseMigrationTypeTokensOnly:
+            BLOG(ledger::LogLevel::LOG_ERROR) << "DB: BAT only migration failed. Initializing without migration." << std::endl;
+            self.dataMigrationFailed = YES;
+            self.migrationType = BATLedgerDatabaseMigrationTypeNone;
+            [self resetRewardsDatabase];
+            // attempt initialize without migrating at all
+            [self initializeLedgerService:NO];
+            return;
+          default:
+            break;
+        }
+      }
+    }
+    self.initializationResult = static_cast<BATResult>(result);
+    for (BATBraveLedgerObserver *observer in [self.observers copy]) {
+      if (observer.walletInitalized) {
+        observer.walletInitalized(self.initializationResult);
+      }
+    }
+  });
+}
+
+- (BOOL)databaseNeedsMigration
+{
+  // Check if we even have a DB to migrate
+  if (!DataController.defaultStoreExists) {
+    return NO;
+  }
+  // Have we set the pref saying ledger has alaready initialized successfully?
+  if ([self.prefs[kMigrationSucceeded] boolValue]) {
+    return NO;
+  }
+  // Can we check the DB just in case we've
+  if (!rewardsDatabase) {
+    return YES;
+  }
+  // Check the meta table
+  auto transaction = ledger::DBTransaction::New();
+  const auto command = ledger::DBCommand::New();
+  command->type = ledger::DBCommand::Type::READ;
+  command->command = "SELECT value FROM meta WHERE key = 'version';";
+  command->record_bindings = { ledger::DBCommand::RecordBindingType::STRING_TYPE };
+  transaction->commands.push_back(command->Clone());
+  
+  auto response = ledger::DBCommandResponse::New();
+  rewardsDatabase->RunTransaction(std::move(transaction), response.get());
+  
+  // Failed to even run the check, table probably doesn't exist
+  if (response->status != ledger::DBCommandResponse::Status::RESPONSE_OK) {
+    return YES;
+  }
+  
+  const auto record = std::move(response->result->get_records());
+  // Meta table exists, but the version key doesn't exist?
+  if (record.empty() || record.front()->fields.empty()) {
+    return YES;
+  }
+  
+  // Field exists so migration has happened already, but somehow the flag wasn't
+  // saved.
+  self.prefs[kMigrationSucceeded] = @(YES);
+  [self savePrefs];
+  
+  return NO;
+}
+
+- (NSString *)rewardsDatabasePath
+{
+  return [self.storagePath stringByAppendingPathComponent:@"Rewards.db"];
+}
+
+- (void)resetRewardsDatabase
+{
+  delete rewardsDatabase;
```

Comment:
Should set `rewardsDatabase` to nullptr in the constructor before the `if (!self.prefs)` call because calling `delete` on an uninitialized pointer will be UB.

Reply:
Good point, do you think this is still necessary given I am reallocing it right below?

Reply:
Should be set to null in the `constructor`. It's not necessary below :)

Reply:
Fixed